README.md
# ms-rules

> Simple rule engine written in TS

[![npm](https://img.shields.io/npm/v/ms-rules.svg?style=flat-square)](https://www.npmjs.com/package/ms-rules) [![](https://img.shields.io/npm/v/ms-rules/beta?style=flat-square)](https://www.npmjs.com/package/ms-rules)

## Installation

```shell
npm i ms-rules
```

## Example

You can find full examples from [examples folder](https://github.com/MamoruDS/ms-rules/tree/main/examples).

### declare new rule engine

-   JS

    ```javascript
    const RuleEngine = require('ms-rules').RuleEngine

    const ACTIONS = ['ACTION_1', 'ACTION_2']

    // Define your function map
    const fnMap = {
        type1: {
            fn: (input, value) => {
                if (/* some conditions */) {
                    return true
                }
                return false
            },
        },
        type2: {
            fn: (input, value) => {
                if (/* some conditions */) {
                    return true
                }
                return false
            },
            valType: 'string', // optional value type check
            inputType: (input) => {
                // optional input type check
                return (/* some conditions */)
            }
        },
    }

    // Declare your rule engine
    const rule = new RuleEngine(fnMap, 'FINAL_ACTION')
    rule.actions = ACTIONS // optional action name checking
    ```

-   TS

    ```typescript
    import { RuleEngine } from 'ms-rules'

    const ACTIONS = ['ACTION_1', 'ACTION_2'] as const

    type RuleType = 'type1' | 'type2'
    type RuleAction = typeof ACTIONS[number]
    // type RuleAction = 'ACTION_1' | 'ACTION_2' // alternaitve

    // Define your function map
    const fnMap: RuleFnMap<RuleType> = {
        type1: {
            fn: (input: number, value: number) => {
                return input == (value || 0)
            },
        },
        type2: {
            fn: (input: { message: string }) => {
                return false
            },
            valType: 'string', // optional value type check
            inputType: (input) => {
                // optional input type check
                return (
                    typeof input == 'object' &&
                    typeof input['message'] == 'string'
                )
            },
        },
    }

    // Declare your rule engine
    const rule = new RuleEngine<RuleAction, RuleType>(fnMap, 'ACTION_2')
    rule.actions = ACTIONS // optional action name checking
    ```

### rule register

```javascript
rule.load({
    type: 'OR',
    action: 'ACTION_1',
    list: [
        {
            type: 'type1',
            value: -1,
        },
        {
            type: 'AND',
            list: [
                {
                    type: 'type2',
                    value: 'foo',
                },
                {
                    type: 'type2',
                    value: 'bar',
                },
            ],
        },
        {
            type: 'NOT',
            list: [
                {
                    type: 'type2',
                    value: 'something',
                },
            ],
        },
    ],
})
```

### test your rules

```javascript
rule.exec({
    message: 'something contains word_2',
})

rule.exec(500)
```

### Show plot in terminal

```javascript
rule.exec(
    {
        id: '12345671890',
        message: 'something contains word_2',
    },
    true,
    false
)
```

<p align="center">
  <img width="550px" src="https://raw.githubusercontent.com/MamoruDS/ms-rules/dev/static/Screen%20Shot%202020-12-13%20at%2011.43.22%20AM.png">
</p>

## Changelog

[link](https://github.com/MamoruDS/ms-rules/blob/main/CHANGELOG.md) of changelogs.

## License

MIT © MamoruDS


CHANGELOG.md
# Changelog

## 1.0.2

### Added:
- new property `actions` in `RuleEngine`
- `getter` / `setter` for properties `actions` and `final`
- strict action name checking in `RuleEngine._test`

## 1.0.1

### Added

-   export `RuleCheckError`
-   added global config `silentMode`
-   show rule match plot in terminal
    ![](https://raw.githubusercontent.com/MamoruDS/ms-rules/dev/static/Screen%20Shot%202020-12-13%20at%2011.43.22%20AM.png)

### Changed

-   new parameters `haltOnceFailed` and `log` to method `RuleEngine.load`
-   new parameters `lazyMatch` and `showPlot` to method `RuleEngine.exec`
-   modified return of method `RuleEngine.load` from `void` to `errMsgs as string[]`
-   modified return of method `RuleEngine.exec` from `Action as string` to
    ```typescript
    {
        matches: {
            [key in MatchKey]: key extends 'match'
                ? boolean
                : key extends 'action'
                ? A
                : Matches<A>
        }
        action: A // final action, also origin return of method `RuleEngine.exec`
    }
    // example:
    {
        '0': {
            '0': { match: true },
            '1': {
                '1': {
                    match: true,
                },
                match: true,
            },
            match: true,
            action: 'ACTION_0A',
        },
        '1': { match: true, action: 'ACTION_00' },
    }
    ```

## 1.0.0


src/line.ts
import * as CONST from './constant'
import { RuleLineStatus } from './types'

class RuleLine<A extends string, T extends string> {
    private type: T | CONST.TYPE_RESERVED
    private action: A
    private value: any
    private indent: number
    private size: number = 4
    private status: RuleLineStatus = 0
    constructor(
        type: T | CONST.TYPE_RESERVED,
        action: A,
        indent: number,
        value?: any
    ) {
        this.type = type
        this.value = value
        this.action = action
        this.indent = indent
    }
    private _prefix(): string {
        return ' '.repeat(this.size).repeat(this.indent)
    }
    private _type(): string {
        const { cs, cp } = this._color()
        return this._highlight(` ${this.type} `, cp, cs)
    }
    private _action(): string {
        const { cs, cp, tint } = this._color()
        return this.action
            ? this._highlight(` ${this.action}`, undefined, tint, '1')
            : ''
    }
    private _value(): string {
        const { cs, cp } = this._color()
        return this.value
            ? this._highlight(` ${this.value} `, '0', undefined, undefined)
            : ''
    }
    private _highlight(
        input: string,
        bg?: string,
        fg?: string,
        style?: string
    ) {
        return `${bg ? `\x1b[4${bg}m` : ''}${fg ? `\x1b[3${fg}m` : ''}${
            style ? `\x1b[${style}m` : ''
        }${input}\x1b[0m`
    }
    private _color() {
        const s = this.status
        return {
            cp:
                s == -1
                    ? '1' // -1
                    : s == 0
                    ? undefined // 0
                    : s == 1
                    ? '4' // 1
                    : undefined, //2
            cs:
                s == -1
                    ? undefined // -1
                    : s == 0
                    ? undefined // 0
                    : s == 1
                    ? undefined // 1
                    : '4', // 2
            tint:
                s == -1
                    ? '1' // -1
                    : s == 0
                    ? undefined // 0
                    : s == 1
                    ? '4' // 1
                    : '4', // 2
        }
    }
    update(status: RuleLineStatus): void {
        this.status = status
    }
    toString(): string {
        return `${this._prefix()}${this._type()}${this._value()}${this._action()}`
    }
}

export { RuleLine }


src/main.ts
import * as CONST from './constant'
import {
    Rule,
    RuleFn,
    RuleFnMap,
    Path,
    Matches,
    RuleLineStatus,
    RuleCheckError,
} from './types'
import { RuleLine } from './line'
import { println } from './utils'

const config = {
    silentMode: false,
}

class RuleEngine<A extends string, T extends string> {
    private _ruleFnMap: RuleFnMap<T>
    private _rules: Rule<A, T>[]
    private _final: A
    private _actions: A[] = []
    constructor(fnMap: RuleFnMap<T>, finalAction: A) {
        this._ruleFnMap = fnMap
        this._rules = []
        this._final = finalAction
    }
    get final(): A {
        return this._final
    }
    set final(action: A) {
        this._final = action
    }
    get actions(): A[] {
        return this._actions
    }
    set actions(actions: A[]) {
        this._actions = actions
    }
    private _check(
        rule: Rule<A, T>,
        path: Path[] = []
    ): {
        error: boolean
        errorMsg?: string
    } {
        let res = {
            error: false,
            errorMsg: undefined,
        }
        const e = (msg: string, format: boolean = true) => {
            const errorMsg = format
                ? msg +
                  ` at line:${path[0].lineNr}; path: [${[...path]
                      .reverse()
                      .map((p) => {
                          return p.type
                      })
                      .join('/')}]`
                : msg
            path.shift()
            throw new RuleCheckError(errorMsg)
        }
        const isSubRule = (): boolean => {
            for (const p of [...path.slice(1)].reverse()) {
                if (CONST.TYPE_GROUP.indexOf(p.type) != -1) {
                    return true
                }
            }
            return false
        }
        path.unshift({
            lineNr: this._rules.length + 1,
            type: rule.type,
            action: rule.action,
        })
        if (!isSubRule() && typeof rule.action == 'undefined') {
            // TODO:
            e(`Missing property 'action' of rule`)
        } else if (typeof rule.action != 'undefined') {
            if (
                Array.isArray(this._actions) &&
                this._actions.length > 0 &&
                this._actions.indexOf(rule.action) == -1
            ) {
                e(`Unknown action '${rule.action}'`)
            }
        }
        if (CONST.TYPE_SINGLE.indexOf(rule.type) != -1) {
        } else if (CONST.TYPE_GROUP.indexOf(rule.type) != -1) {
            if (typeof rule.list == 'undefined') {
                e(`Missing property 'list' of rule group`)
            }
            if (
                rule.list.length >
                CONST.LEN_LIMIT[rule.type as CONST.TYPE_GROUP]
            ) {
                e(`Number of rules in group exceeds the limit`)
            }
            for (const r of rule.list) {
                const res = this._check(r, path)
                if (res.error) {
                    e(res.errorMsg, false)
                }
            }
        } else if (Object.keys(this._ruleFnMap).indexOf(rule.type) != -1) {
            const valType = this._ruleFnMap[rule.type as T].valType || false
            if (
                valType &&
                (valType != 'array'
                    ? typeof rule.value != valType
                    : !Array.isArray(rule.value))
            ) {
                e(
                    `Invalid rule value: <${valType}> is required but given '${
                        rule.value
                    }' as <${typeof rule.value}>`
                )
            }
        } else {
            e(`Unknown type '${rule.type}'`)
        }
        path.shift()
        return res
    }
    load(rule: Rule<A, T>, haltOnceFailed?: boolean, log?: true): string[] {
        const errs: string[] = []
        try {
            this._check(rule)
        } catch (e) {
            if (e instanceof RuleCheckError) {
                if (haltOnceFailed) {
                    throw e
                }
                if (log) {
                    errs.push(e.message)
                } else {
                    println(e.message, 'error')
                }
            } else {
                throw e
            }
        }
        this._rules.push(rule)
        return errs
    }
    private _test(
        rule: Rule<A, T>,
        input: any,
        index: string[],
        matches: Matches<A>,
        actions: A[],
        show?: RuleLine<A, T>[]
    ) {
        let match: number
        let line: RuleLine<A, T>
        if (show) {
            line = new RuleLine<A, T>(
                rule.type,
                rule.action,
                index.length - 1,
                rule.value
            )
            show.push(line)
        }
        if (CONST.TYPE_RESERVED.indexOf(rule.type) != -1) {
            const update = (m: number) => {
                match = CONST.RULE_FN_MAP_RESERVED[
                    rule.type as CONST.TYPE_RESERVED
                ]?.fn({
                    lhs: match,
                    rhs: m,
                })
                    ? 1
                    : 0
            }
            if (CONST.TYPE_GROUP.indexOf(rule.type) != -1) {
                for (const i in rule['list']) {
                    const m = this._test(
                        rule['list'][i],
                        input,
                        [...index, i],
                        matches,
                        actions,
                        show
                    )
                    update(m)
                }
            } else {
                // TODO:
                update(0)
            }
        }
        if (Object.keys(this._ruleFnMap).indexOf(rule.type) != -1) {
            const ruleFn = this._ruleFnMap[rule.type as T]
            const inType = ruleFn.inputType
            if (inType && typeof inType == 'function') {
                if (!inType(input)) {
                    match = 0
                } else {
                    match = ruleFn?.fn(input, rule.value) ? 1 : 0
                }
            } else if (
                inType &&
                (inType != 'array'
                    ? typeof input != inType
                    : !Array.isArray(input))
            ) {
                match = 0
            } else {
                match = ruleFn?.fn(input, rule.value) ? 1 : 0
            }
        }
        if (match == 1) {
            const _index = [...index]
            let t = matches
            while (_index.length != 0) {
                const i = parseInt(_index.shift())
                if (typeof t[i] == 'undefined') {
                    t[i] = {
                        match: false,
                    }
                }
                t = t[i]
                if (_index.length == 0) {
                    t['match'] = true
                    if (typeof rule.action != 'undefined') {
                        t['action'] = rule.action
                        actions.push(rule.action)
                    }
                }
            }
        }
        if (show) {
            const _index = [...index]
            let t = matches
            let m: RuleLineStatus = match ? 1 : 0
            while (_index.length != 0) {
                const i = parseInt(_index.shift())
                if (typeof t[i] == 'undefined') {
                    line.update(m)
                    break
                }
                t = t[i]
                if (_index.length == 0) {
                    const sub = Object.keys(t).filter((key) => {
                        return key != 'match' && key != 'action'
                    })
                    if (sub.length == 0) {
                        line.update(m)
                        break
                    } else {
                        let matched = false
                        for (const key of sub) {
                            matched = matched || t[key]['match']
                        }
                        if (matched) {
                            line.update(m == 0 ? -1 : 2)
                        }
                    }
                }
            }
        }
        return match
    }
    exec(
        input: any,
        showPlot: boolean = false,
        lazyMatch: boolean = true
    ): {
        matches: Matches<A>
        action: A
    } {
        const rules: Rule<A, T>[] = [
            ...this._rules,
            {
                type: 'FINAL',
                action: this._final,
            },
        ]
        const matches = {} as Matches<A>
        const actions: A[] = []
        const lines: RuleLine<A, T>[] = showPlot ? [] : undefined
        if (showPlot && lazyMatch) {
            println(
                `lazyMatch will be disabled since showPlot has been enabled.`,
                'info'
            )
        }
        for (const i in rules) {
            this._test(rules[i], input, [i], matches, actions, lines)
            if (!showPlot && lazyMatch && actions.length > 0) {
                break
            }
        }
        if (showPlot) {
            println(
                lines
                    .map((line) => {
                        return line.toString()
                    })
                    .join('\n')
            )
        }
        return {
            matches,
            action: actions.shift(),
        }
    }
}

export { config }
export { Rule, RuleFn, RuleFnMap }
export { RuleEngine, RuleCheckError }


src/types.ts
import * as CONST from './constant'

interface Rule<A extends string, T extends string> {
    type: T | CONST.TYPE_RESERVED
    list?: Rule<A, T>[]
    value?: any
    comment?: string
    action?: A
}

type RuleFn = (input: any, value?: any) => boolean

type RuleFnMap<T extends string> = {
    [key in T]: {
        fn: RuleFn
        valType?:
            | 'undefined'
            | 'object'
            | 'array'
            | 'boolean'
            | 'number'
            | 'string'
            | 'function'
        inputType?:
            | 'undefined'
            | 'object'
            | 'array'
            | 'boolean'
            | 'number'
            | 'string'
            | 'function'
            | ((input: any) => boolean)
    }
}

type Path = {
    lineNr: number
    type: string
    action: string
}

class RuleCheckError extends Error {
    constructor(message: string) {
        super(message)
    }
}

type _Decimal = '1' | '2' | '3' | '4' | '5' | '6' | '7' | '8' | '9'

type Decimal = '0' | _Decimal

type MatchKey =
    | 'match'
    | 'action'
    | `${Decimal}` // 0 - 9
    | `${_Decimal}${Decimal}` // 10 - 99

type Matches<A> = {
    [key in MatchKey]: key extends 'match'
        ? boolean
        : key extends 'action'
        ? A
        : Matches<A>
}

type RuleLineStatus = -1 | 0 | 1 | 2

export { Rule, RuleFn, RuleFnMap, Path, Matches, RuleLineStatus }
export { RuleCheckError }


src/constant.ts
import { RuleFnMap } from './types'

type TYPE_SINGLE = 'FINAL'
type TYPE_GROUP = 'AND' | 'OR' | 'NOT'

const TYPE_NOT = 'NOT'
const TYPE_SINGLE = ['FINAL']
const TYPE_GROUP = ['AND', 'OR', 'NOT']
const LEN_LIMIT: {
    [key in TYPE_GROUP]: number
} = {
    AND: Infinity,
    OR: Infinity,
    NOT: 1,
}

const TYPE_RESERVED = [...TYPE_SINGLE, ...TYPE_GROUP]
type TYPE_RESERVED = TYPE_SINGLE | TYPE_GROUP

const RULE_FN_MAP_RESERVED: RuleFnMap<TYPE_RESERVED> = {
    AND: {
        fn: (input: { lhs: number; rhs: number }) => {
            if (typeof input['lhs'] == 'undefined') {
                return Boolean(input['rhs'])
            } else {
                return Boolean(input['lhs'] & input['rhs'])
            }
        },
        valType: 'undefined',
    },
    OR: {
        fn: (input: { lhs: number; rhs: number }) => {
            if (typeof input['lhs'] == 'undefined') {
                return Boolean(input['rhs'])
            } else {
                return Boolean(input['lhs'] | input['rhs'])
            }
        },
        valType: 'undefined',
    },
    NOT: {
        fn: (input: { lhs: number; rhs: number }) => {
            return !Boolean(input['rhs'])
        },
        valType: 'undefined',
    },
    FINAL: {
        fn: () => {
            return true
        },
        valType: 'undefined',
    },
}

export { TYPE_NOT, TYPE_SINGLE, TYPE_GROUP, TYPE_RESERVED }

export { LEN_LIMIT, RULE_FN_MAP_RESERVED }


examples/messageFilter.ts
import { RuleEngine } from '../src/main'
// import { RuleEngine } from 'ms-rules'
import { RuleFnMap } from '../src/main'
// import { RuleFnMap } from 'ms-rules'

const actions = ['PUSH', 'DISCARD'] as const

type RuleType = 'id' | 'keyword'
type RuleAction = typeof actions[number]

const _: RuleAction = 'PUSH'

// ## Define your function map
const fnMap: RuleFnMap<RuleType> = {
    id: {
        fn: (input: { id: string }, value: string) => {
            return input?.id == value
        },
    },
    keyword: {
        fn: (input: { message: string }, value: string) => {
            return (input?.message ?? '').indexOf(value) != -1
        },
    },
}

// ## Declare your rule engine
const rule = new RuleEngine<RuleAction, RuleType>(fnMap, 'PUSH')
rule.actions = [...actions]

// ## Load your rules
rule.load({
    type: 'AND',
    action: 'DISCARD',
    list: [
        {
            type: 'id',
            value: '1234567890',
        },
        {
            type: 'OR',
            list: [
                {
                    type: 'keyword',
                    value: 'word_1',
                },
                {
                    type: 'keyword',
                    value: 'word_2',
                },
                {
                    type: 'keyword',
                    value: 'word_3',
                },
                {
                    type: 'NOT',
                    list: [
                        {
                            type: 'keyword',
                            value: 'blocked_word',
                        },
                    ],
                },
            ],
        },
    ],
})

// ## Test your rules
let action: string

action = rule.exec({
    id: '1234567890',
    message: 'something contains word_2',
}).action
console.log(action) // DISCARD

action = rule.exec({
    id: '1234567890',
    message: 'bonjour',
}).action
console.log(action) // PUSH

action = rule.exec(
    {
        id: '0987654321',
        message: 'something contains word_3',
    },
    false // disable lazyMatch (optional)
).action
console.log(action) // PUSH

// ### Show match plot in terminal
rule.exec(
    {
        id: '12345671890',
        message: 'something contains word_2',
    },
    true,
    false
)


examples/messageFilter.js
const RuleEngine = require('../dist/main').RuleEngine
// const RuleEngine = require('ms-rules').RuleEngine

const actions = ['PUSH', 'DISCARD']

// ## Define your function map
const fnMap = {
    id: {
        fn: (input, value) => {
            if (typeof input == 'object' && !Array.isArray(input)) {
                return input['id'] == value
            }
            return false
        },
    },
    keyword: {
        fn: (input, value) => {
            if (
                typeof input == 'object' &&
                typeof input['message'] == 'string'
            ) {
                return input['message'].indexOf(value) != -1
            }
            return false
        },
    },
}

// ## Declare your rule engine
const rule = new RuleEngine(fnMap, 'PUSH')
rule.actions = actions

// ## Load your rules
rule.load({
    type: 'AND',
    action: 'DISCARD',
    list: [
        {
            type: 'id',
            value: '1234567890',
        },
        {
            type: 'OR',
            list: [
                {
                    type: 'keyword',
                    value: 'word_1',
                },
                {
                    type: 'keyword',
                    value: 'word_2',
                },
                {
                    type: 'keyword',
                    value: 'word_3',
                },
                {
                    type: 'NOT',
                    list: [
                        {
                            type: 'keyword',
                            value: 'blocked_word',
                        },
                    ],
                },
            ],
        },
    ],
})

// ## Test your rules
let action

action = rule.exec({
    id: '1234567890',
    message: 'something contains word_2',
}).action
console.log(action) // DISCARD

action = rule.exec({
    id: '1234567890',
    message: 'bonjour',
}).action
console.log(action) // PUSH

action = rule.exec(
    {
        id: '0987654321',
        message: 'something contains word_3',
    },
    false // disable lazyMatch (optional)
).action
console.log(action) // PUSH

// ### Show match plot in terminal
rule.exec(
    {
        id: '12345671890',
        message: 'something contains word_2',
    },
    true,
    false
)
